src/destructors.md

```diff
@@ -37,57 +33,344 @@ impl Drop for ShowOnDrop {
     }
 }
 
+let mut overwritten = ShowOnDrop("drops when overwritten");
+overwritten = ShowOnDrop("drops when scope ends");
+
+let tuple = (ShowOnDrop("Tuple first"), ShowOnDrop("Tuple second"));
+
+let moved;
+// No destructor run on assignment.
+moved = ShowOnDrop("Drops when moved");
+// drops now, but is then uninitialized
+moved;
+let uninitialized: ShowOnDrop;
+// Only the first element is dropped
+let mut partially_initialized: (ShowOnDrop, ShowOnDrop);
+partially_initialized.0 = ShowOnDrop("Partial tuple first");
+```
+
+## Drop scopes
+
+Each variable or temporary is associated to a *drop scope*. When control flow
+leaves a drop scope all variables associated to that scope are dropped in
+reverse order of declaration (for variables) or creation (for temporaries).
+
+Drop scopes are determined after replacing [`for`], [`if let`] and
+[`while let`] expressions with the equivalent expressions using [`match`].
+Overloaded operators are not distinguished from built-in operators and [binding
+modes] are not considered.
+
+Given a function, or closure, there are drop scopes for:
+
+* The entire function
+* Each [statement]
+* Each [expression]
+* Each block, including the function body
+    * [Block expressions] the expression and block scopes are the same scope.
```

Comment:
This doesn't read well; separate "block expressions" and "the expression" somehow?